stringfuzz/generators/equality.py

```diff
@@ -0,0 +1,107 @@
+import random
+
+from stringfuzz.scanner import ALPHABET
+from stringfuzz.smt import *
+
+__all__ = [
+    'equality',
+]
+
+def random_string(size):
+    return ''.join(random.choice(ALPHABET) for i in range(size))
+
+def get_length(max_length, randomise):
+    if randomise is False:
+        return max_length
+    return random.randint(0, max_length)
+
+def concat_terms(terms):
+
+    # initialise result to the last term (i.e. first in reversed list)
+    reversed_terms = reversed(terms)
+    result         = next(reversed_terms)
+
+    # keep concatenating preceding terms to the result
+    for term in reversed_terms:
+        result = smt_concat(term, result)
+
+    return result
+
+def randomly_add_infix(probability):
+    return random.random() < probability
+
+def make_equality(num_expressions, num_terms, prefix_length, suffix_length, add_infixes, infix_length, randomise_lengths, infix_probability):
+
+    # check args
+    if num_expressions < 1:
+        raise ValueError('the number of expressions must be at least 1')
+
+    if num_terms < 2:
+        raise ValueError('the number of terms per expression must be at least 2')
+
+    if infix_probability < 0.0 or 1.0 < infix_probability:
+        raise ValueError('the probability of infixes must be between 0.0 and 1.0')
+
+    # result values
+    expressions = []
+    variables   = []
+    heads       = []
+
+    # create expressions
+    for i in range(num_expressions):
+
+        # create head, prefix, and suffix
+        head   = smt_new_var()
+        prefix = smt_str_lit(random_string(get_length(prefix_length, randomise_lengths)))
+        suffix = smt_str_lit(random_string(get_length(suffix_length, randomise_lengths)))
+
+        # keep track of new variables
+        new_variables = [head]
+        heads.append(head)
+
+        # create middle
+        middle = []
+        for i in range(num_terms - 2):
+
+            # if infixes are enabled, add them with the given probability
+            if add_infixes is True and randomly_add_infix(infix_probability) is True:
+                new_term = smt_str_lit(random_string(get_length(infix_length, randomise_lengths)))
+
+            # otherwise, just add variables
+            else:
+                new_term = smt_new_var()
+                new_variables.append(new_term)
+
+            middle.append(new_term)
+
+        # compose full expression
+        terms    = [prefix] + middle + [suffix]
+        concat   = concat_terms(terms)
+        equality = smt_assert(smt_equal(head, concat))
+
+        # remember variables and expressions
+        variables += new_variables
+        expressions.append(equality)
+
+    # finally, create equality between all expressions
```

Comment:
What happens if instead you make it a cycle? Will some solvers have an easier/harder time? Maybe can set this as a command line argument

Reply:
What do you mean by a cycle? Also, maybe to make it easier I can just set all the concats to be equal to the same variable. Do you think that might be cleaner and simpler?

Reply:
The equality between all expressions is set like this right now (right?):
e1 = e2
e1 = e3
...
e1 = en

Would solvers find a difference between solving that sort of equality versus solving one like (this is what I meant by cycle, but that's probably the wrong word; maybe chain is better):
e1 = e2
e2 = e3
...
en-1 = en

Reply:
Of course ideally I'd expect solvers to rewrite that as just one variable, but yeah, in practice they might treat those differently. I'll go with the one-variable solution.